File: BrokenScripts/BuildQueue.cs
using UnityEngine;
using System.Collections.Generic;

public class BuildQueue : MonoBehaviour {
	
	public List<BuildOrder> queue;
	public int playerId = 0;
	
	//private Player player;
	
	// Use this for initialization
	void Start () {
		//player = Player.GetPlayer(playerId);
	}
	
	void Advance() {
		if(queue.Count <= 0) return;
		
		// look at the next thing in queue
		BuildOrder next = queue[0];
		
		// if we have enough money for it, submit the build order!
		
		// need to know where the hell to send the message, though!
		
	}
	
	private int GetTotalValue()
	{
		int total = 0;
		foreach(BuildOrder b in queue)
		{
			total += b.cost;
		}
		return total;
	}
}

public class BuildOrder
{
	public Texture2D icon;
	public string description;
	public int cost;
	public GameObject sender;	// wot sent the order
	public string buildMessage; // the message to use in SendMessage when the queue decides to build this order
}

File: Assets/Scripts/BuildQueue.js
import System.Collections.Generic;

public var queue = new List.<BuildOrder>();
	
private var player : Player;
	
function Start () {
	player = GetComponent(Player);
}

function Advance() {
	var next : BuildOrder;
	var toRemove = 0;
	
	// look at the next thing in queue
	for(next in queue)
	{
		// if we have enough money for it, submit the build order!
		if(next.cost > player.cash) break;
		
		// need to know where the hell to send the message, though!
		next.sender.SendMessage(next.buildMessage, next.buildArgs);
	
		//FIXME : for now, I am not manually deducting the cost
		//  because the commands that get called end up deducting the cost
		//  but it's probably smartest to move that functionality here
		//player.cash -= next.cost;
		Debug.Log("build order completed! cost $" + next.cost + "  --  ");
		
		toRemove += 1;
	}
	
	if(toRemove > 0)
	{
		queue.RemoveRange(0,toRemove);
	}
}


function OnSubmit(order : BuildOrder)
{
	queue.Add(order);
	
	// immediately evaluate if we can build this or not
	Advance();
}

function Submit(cost : int, description : String, icon : Texture2D, sender : GameObject, buildMessage : String)
{
	// this is kind of dumb, but keeps other scripts from having to know about BuildOrder
	var order = new BuildOrder();
	order.cost = cost;
	order.description = description;
	order.icon = icon;
	order.sender = sender;
	order.buildMessage = buildMessage;
	
	queue.Add(order);
	
	// immediately evaluate if we can build this or not
	Advance();
}


function GetTotalValue() : int
{
	var total = 0;
	for(var b : BuildOrder in queue)
	{
		total += b.cost;
	}
	return total;
}
